### PR TITLE
[WIP] Embeddings retrain

### PR DESCRIPTION
**Pull request by pcuenca**
I tried to adapt the code from https://colab.research.google.com/drive/1g_pt2Rc3bv6H6qchvGHD-BpgF-Pt4vrC#scrollTo=WWHlukuvIpXb, but it's hard due to our split and partitioning mechanism.

It currently works (I think) for `adafactor`, but not for `adam` or `distributed_shampoo`.

The idea we use here is to replace the optimizer with a `multi_transform` that applies different optimizations to frozen vs trainable parameters. I believe that `optax.masked()` does something similar. I also explored the approach to create a new "frozen" split to zero those gradients, but I had trouble with axis when pjitting the state.

### CODE REVIEW

**Comment by borisdayma** [comment]
This is really cool! I love the notebook!
Do we need to also have the option to init missing parameters in the training script?

**Comment by pcuenca** [comment]
> Do we need to also have the option to init missing parameters in the training script?

We could potentially do that, yes, but I thought it would be as easy to save the checkpoint with the newly-initialized layers as with those layers removed. If we do it, what would be a principled way to know that there are missing layers? Just look at the `_missing_keys` property?

**Comment by borisdayma** [comment]
Oh yes you're right. Can you save the checkpoint somewhere on wandb?

**Comment by pcuenca** [comment]
Oh, sure, yes. I pushed it here: https://wandb.ai/dalle-mini/dalle-mini/runs/2iz3rx4n/artifacts
It comes from `dalle-mini/dalle-mini/mini-1:v0`.

File: tools/train/embeddings_retrain_preparation.ipynb [hunk 1]
```diff
@@ -0,0 +1,1030 @@
+{
```
Comment (borisdayma) [inline tools/train/embeddings_retrain_preparation.ipynb:1]: <p>I checked in more details.</p><p> </p><p>The weights we want to reinit are:</p><ul><li>params.lm_head.* (kernel + bias if present)</li><li>params.model.decoder.embed_tokens.*</li><li>params.model.decoder.embed_positions.*</li><li>params.model.decoder.final_ln.*</li><li>params.model.decoder.layernorm_embedding.*</li></ul><p> </p><p>For the encoder I would not reinit because it's on the text only.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/borisdayma/dalle-mini/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (pcuenca) [reply tools/train/embeddings_retrain_preparation.ipynb:1]: Thanks, I'll update it! I did not reinit the encoder layers but was not sure about `lm_head`.

**Comment by pcuenca** [comment]
Superseded by #291, #293.
